File: estructura-obligatorio/java/tads/AVLSet.java
package tads;

public class AVLSet<T extends Comparable<T>>{
    private AVL<T> tree;
    public int size(){
      //TODO
      return 0;
    }

    public boolean contains(T data){
      // returns true if data is in the set
      //TODO
      return false;
    }

    public void clear(){
      //TODO
    }

    public void add(T data){
      // add data to the set
      //TODO
    }

    public void remove(T data){
      // remove data from the set
      //TODO
    }

    public void join(AVLSet<T> other){
      // var it = other.tree.iterator();
      // while(it.hasNext()){
      //   T data = it.next();
      //   this.add(data);
      // }
      for (T data : other.tree) {
        this.add(data);
      }
    }

    public AVLSet<T> intersect(AVLSet<T> other){
      // returns the intersection of this set and other set
      //TODO
      return null;
    }

    public AVLSet<T> difference(AVLSet<T> other){
      // returns the difference of this set and other set
      //TODO
      return null;
    }
  }

File: estructura-obligatorio/c++/tads/AVLSet.cpp
#pragma once

#include "AVL.cpp"

template <class T>
class AVLSet
{
public:
    int size()
    {
        return tree.size();
    }

    bool contains(T data)
    {
        return tree.contains(data);
    }

    void clear()
    {
        tree.clear();
    }

    //post: el dato existe en el Set
    void add(T data)
    {
        if (!tree.contains(data))
        {
            tree.add(data);
        }
    }

    void remove(T data)
    {
        tree.remove(data);
    }

    void join(AVLSet<T> *other)
    {
        auto iter = other->iterator();
        while (iter->hasNext())
        {
            T otherData = iter->next();
            this->add(otherData);
        }
        // returns the union of this set and other set
        // TODO
    }

    void intersect(AVLSet<T> *other)
    {
        auto iter = this->iterator();
        while (iter->hasNext())
        {
            T data = iter->next();
            if (!other->contains(data))
            {
                this->remove(data);
            }
        }
    }

    void difference(AVLSet<T> *other)
    {
        auto iter = other->iterator();
        while (iter->hasNext())
        {
            T otherData = iter->next();
            if (this->contains(otherData))
            {
                this->remove(otherData);
            }
            else
            {
                this->add(otherData);
            }
        }
    }

    Iterator<T> *iterator()
    {
        return tree.iterator();
    }

private:
    AVL<T> tree = AVL<T>();
};
